### PR TITLE
[experimental_v1.0] New PrivacyEngine prototype

### PR DESCRIPTION
**Pull request by ffuuugor**
Disclaimer: it doesn't work yet. I've tested the design on some simple cases (MNIST example) to check it's valid, but all the tests will break and many features are missing

## Highlights

We embrace the concept of isolated decoupled modules with as few responsibilities as possible. They're all tied together into a simple API by PrivacyEngine.

We have:
- `GradSampleModule`. Already there. Responsible for computing per sample gradients and nothing else. Can be used outside of DP context.
- `DPOptimizer`. Wrapper around regular optimizer. Performs clipping and adds noise, then calls for underlying optimizer's `step`. Takes most of the core responsibilities previously owned by PrivacyEngine. Provides hooks to do accounting.
- `DPDataLoader`. Wrapper around regular data loader, but takes control over sampling. The rest is normal DataLoader. Side benefit: it also controls `collate` method, so we can handle 0-sized batches much more elegantly than today.

PrivacyEngine now becomes very minimalistic object, mostly providing a one-stop interface for the three classes above:
```
model = MyModule()
optimizer = optim.SGD(model.parameters(), lr=lr, momentum=0)
train_loader = DataLoader(<...>)

privacy_engine = PrivacyEngine()

model, optimizer, train_loader = privacy_engine.prepare(
    module=model,
    optimizer=optimizer,
    data_loader=train_loader,
    noise_multiplier=1.0,
    max_grad_norm=1.0
)
``` 

Also note the new interface for PrivacyAccountant - it opens up a road to variable noise and sampling rate:

```
	
class RDPAccountant:
    def __init__(self):
        self.steps = []

    def step(self, noise_multiplier, sample_rate):
        self.steps.append((noise_multiplier, sample_rate))

    def get_privacy_spent(self, delta, alphas):
        pass
```

Overall, I'm quite happy with how it's working out so far. There seem to be a clear responsibility split between `GradSampleModule`, `DPOptimizer`, `DPDataLoader` and `RDPAccountant`. There's surprisingly little code required for the basic DP-SGD implementation, which I really like. 
One thing I'm particularly happy about is that gradients are no longer updated in `NormClipper`, but in a more appropriate place.
 
More importantly, these modules can be used independently of each other. `PrivacyEngine` has become essentially a syntactic sugar - it just passes arguments to constructors with almost zero manipulations. It's a good feature to have for beginners, but I expect researches to be able to play around with privacy modules directly, not through privacy engine. More examples of that in the section below on how to do stuff

## How can I do X?

**variable noise**
You call `PrivacyEngine.prepare()` multiple times, changing the amount of noise or clipping threshold. Optimizer and accountant will pick up the change

**variable sampling rate**
You call `PrivacyEngine.prepare()` multiple times, changing the data loader.  Optimizer and accountant will pick up the change

**Different accounting**
You ditch PrivacyEngine and pass your accountant as a callback to `DPOptimizer`

**Different clipping**
You create a subclass of `DPOptimizer` and override `clip_and_accumulate()` method (potentially also constructor if you need extra arguments)

**Different noise**
Not sure why you'd want that, but you override a method in `DPOptimizer`

**I don't care for uniform sampling, give me batches back**
Ditch PrivacyEngine, use `GradSampleModule` and `DPOptimizer`

## What's next 
There's a lot to do, obviously, but I'll highlight some clearly visible gaps:

- Support DDP. I haven't yet looked into how it fits into the picture, but probably a subclass of `DPOptimizer` should do the job
- Implement different clipping strategies. Probably not required, but nice to have, especially in the context of [#196](https://github.com/pytorch/opacus/pull/196)
- Implement `RDPAccountant`
- Support PackedSequence and padded sequences in `DPDataLoader.collate_fn`
- Support `secure_mode`
- Current implementation of `DPOptimizer.zero_grad()` is sloppy and needs work
- DP model validation
- Current version support virtual steps, but there's a way to make them better. In normal pytorch you can just call forward multiple times before calling `optimizer.step` and that would be your virtual step. I think there's a way to do the same in opacus -  then you'll only need `optimizer.virtual_step` for memory management, gradients and noise will be adjusted automatically
- Tests, obviously

### CODE REVIEW

File: opacus/optimizer.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+            p.grad = p.summed_grad + noise
+
+    def scale_grad(self):
+        if self.loss_reduction == "mean":
+            for p in self.params:
+                p.grad /= (self.expected_batch_size * self.accumulated_iterations)
+
+    # TODO: see GradSampleModule.zero_grad()
+    # TODO: actually, not calling zero_grad() after step() does break privacy accounting - add warning?
+    def zero_grad(self, set_to_none: bool = False):
```
[the remark this answers is not in the document]
Comment (ffuuugor) [reply opacus/optimizer.py:90]: wait a second. I started questioning the premise of my #TODO. Does it break accounting?

As long as we reuse the gradient after we've added noise, it should be fine: it can be considered a post-processing, which is fine under DP guarantees. 

It looks to me as if, as long as we don't accumulate raw gradients between steps, we're good.
Is there any flaw in this logic?

File: opacus/privacy_engine.py [hunk 2]
```diff
@@ -1,659 +1,92 @@
...
-        epochs: Optional[float] = None,
-        loss_reduction: str = "mean",
-        poisson: bool = False,
-        **misc_settings,
+    def __init__(self, secure_mode=False):
+        self.accountant = RDPAccountant()
+        self.secure_mode = secure_mode # TODO: actually support it
+
+    # TODO cool name
+    def prepare(
```
[the remark this answers is not in the document]
Comment (ffuuugor) [reply opacus/privacy_engine.py:18]: >I like this too. the question now is, what should be the primary arg?

Huh, good point. I kinda assumed epsilon is always secondary to the noise multiplier, but I now realise it's not a given.
And I think you're right in bringing up engineers vs researchers trade-off here. Working with bare `noise_multiplier` is more flexible (e.g. you can do scheduling or choose your own stopping condition) and intuitive (as long as you understand DP-SGD).
I would argue, that as long as we stick to prioritizing flexibility for researchers, `noise_multiplier` should remain the main option.

That said, I also want to propose an alternative implementation for PrivacyEngine (in addition to the current one).
This new PrivacyEngine would have a fixed privacy budget (accept epsilon instead on a noise multiplier) and halt the training as soon as this budget is reached (unlike the standard one, which would not signal when you reach your stated budget).

I kinda like the idea of `PrivacyEngine` being a thin layer, implementing simple strategies of putting `DPOptimizer` and other stuff together. But also not sure if it's too much and just adds unnecessary complexity - wdyt?

File: opacus/privacy_engine.py [hunk 3]
```diff
@@ -1,659 +1,92 @@
...
-        target_epsilon: Optional[float] = None,
-        epochs: Optional[float] = None,
-        loss_reduction: str = "mean",
-        poisson: bool = False,
-        **misc_settings,
+    def __init__(self, secure_mode=False):
+        self.accountant = RDPAccountant()
+        self.secure_mode = secure_mode # TODO: actually support it
+
+    # TODO cool name
```
[the remark this answers is not in the document]
Comment (ffuuugor) [reply opacus/privacy_engine.py:17]: >Eg, changing sampling rate (of data loader) would also trigger a change to and return a new optimizer.

That's a very good point, yes. Update-like methods are off the table then.

> I would argue they are semantically different, though functionally the same

Do you have any potential use case in mind, where a subclass would want to have different implementations?

Speaking of semantics, imagine the following scenario: you have a private dataset and want to train two separate models on it. Our current API allows this: you can call `prepare` twice, accountant will keep track of privacy expenditure  from both training loops, and as long as both `prepare`s are called with the same dataset, it all make mathematical sense too.
Now, would you call `prepare()` or `reprepare()` here?

File: opacus/privacy_engine.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexandresablayrolles) [reply opacus/privacy_engine.py:18]: I would prefer to have another function that takes as arguments epsilon, delta and number of epochs, computes noise_multiplier and calls prepare() internally (e.g. your `prepare_with_epsilon` proposal). Usually developers want to set epsilon, delta and number of epochs, I would argue that an Engine that just stops is a bit agressive for the end user!

**Review by alexandresablayrolles — CHANGES_REQUESTED** [review]

File: opacus/optimizer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ffuuugor) [reply opacus/optimizer.py:90]: > If you take two steps without zero_grad, in the second one you would have twice as much data points but only add noise that covers one batch worth of data

True, thanks.

>  I think that the closest to Pytorch would be to accumulate in p.grad

The problem with that is PyTorch will add regular non-clipped gradients to the same field. We had to accumulate in `summed_grad` because we always need to override p.grad as the last step.
What if we do
```
    def add_noise(self):
        for p in self.params:
            noise = _generate_noise(<...>)
            p.summed_grad += noise
            p.grad = p.summed_grad
```

?

[the remark this answers is not in the document]
Comment (ffuuugor) [reply opacus/optimizer.py:90]: There's additional complication of scaling - we need to divide the gradients by batch size.
I think the following could work:

we accumulate raw, non-noised, non-scaled, but clipped gradients in `p.clipped_grad`
on every `step()` we do 
```
p.summed_grad += (p.clipped_grad + noise) / (self.expected_batch_size * self.accumulated_iterations)
p.grad = p.summed_grad
del p.clipped_grad
```
